### DISCUSSION TITLE
Installing NVIDIA drivers

### DISCUSSION THREAD

**Discussion by pbrezina** [bootc-dev/bootc#993] [Q&A] [answered]
Hi, I can't find a way to correctly install nvidia drivers.

The packages are installed (from rpmfusion):
```
# rpm -qa | grep nvidia
nvidia-gpu-firmware-20241210-1.fc41.noarch
nvidia-modprobe-565.77-1.fc41.x86_64
xorg-x11-drv-nvidia-cuda-libs-565.77-1.fc41.x86_64
nvidia-persistenced-565.77-1.fc41.x86_64
xorg-x11-drv-nvidia-libs-565.77-1.fc41.x86_64
xorg-x11-drv-nvidia-kmodsrc-565.77-1.fc41.x86_64
nvidia-settings-565.77-1.fc41.x86_64
xorg-x11-drv-nvidia-xorg-libs-565.77-1.fc41.x86_64
xorg-x11-drv-nvidia-565.77-1.fc41.x86_64
akmod-nvidia-565.77-1.fc41.x86_64
xorg-x11-drv-nvidia-cuda-565.77-1.fc41.x86_64
xorg-x11-drv-nvidia-power-565.77-1.fc41.x86_64
```

But the module is not loaded, the system is still using nouveau:
```
# modinfo -F version nvidia
modinfo: ERROR: Module nvidia not found.

# lspci -n -n -k | grep -A 2 -e VGA -e 3D
01:00.0 VGA compatible controller [0300]: NVIDIA Corporation AD107 [GeForce RTX 4060] [10de:2882] (rev a1)
	Subsystem: ASUSTeK Computer Inc. Device [1043:895e]
	Kernel driver in use: nouveau
```

I suppose I need to add kernel arguments to disable nouveau `rd.driver.blacklist=nouveau modprobe.blacklist=nouveau nvidia-drm.modeset=1`  (mentioned at [rpmfusion docs]https://rpmfusion.org/Howto/NVIDIA#Current_GeForce.2FQuadro.2FTesla)), but I can't find a way to do it. Editing `/boot/loader/entries/ostree-1.conf`, does not include it in the grub menu and [bootc docs](https://containers.github.io/bootc/building/kernel-arguments.html) seem to focus only on setting kargs per-image, not post-installation on single machine.

But even if I put the line manually in grub menu, I get `nvidia module missing, falling back to nouveau`.

Thanks.

**Comment by cgwalters** [comment]
If you use the bootc kargs, changing them in a new container image build will affect the system across upgrades. That's most of the point of the interface.

**Reply by pbrezina** [reply]
And is it possible to share the image with multiple hosts and just use different kargs on selected machine?

**Comment by cgwalters** [comment]
You can create per-machine layers derived from a common base, or use kickstart (or bootc install kargs) to configure per-machine kargs.

**Comment by pbrezina** [comment, marked as answer]
So... the problem main problem was that akmods do not work correctly with bootc and the akmods.service is skipped because `ConditionPathExists=!/run/ostree-booted`.

However, the module was not even built during container build. Installation of akmod-nvidia errors out (but the whole dnf transaction completes so it took me a long time to find this):
```
[75/79] Installing akmod-nvidia-3:565.7 100% | 1.3 MiB/s | 91.4 KiB | 00m00s >>> Running post-install scriptlet: akmod-nvidia-3:565.77-1.fc41.x86_64 >>> Non-critical error in post-install scriptlet: akmod-nvidia-3:565.77-1.fc41.x >>> Scriptlet output: >>> Building /usr/src/akmods/nvidia-kmod-565.77-1.fc41.src.rpm for kernel 6.12.5 >>> ERROR: Not to be used as root; start as user or 'akmodsbuild' instead. >>> >>> [RPM] %post(akmod-nvidia-3:565.77-1.fc41.x86_64) scriptlet failed, exit stat 
```

Additionally, installing kmod-nvidia does not help, since this is only a metapackage that pulls in akmod-nvidia anyway.

I found this workaround:
```
# Install NVIDIA drivers
RUN dnf install -y --allowerasing \
        akmod-nvidia \
        xorg-x11-drv-nvidia \
        xorg-x11-drv-nvidia-cuda \
        ${NULL}

# Rebuild akmods
RUN akmods --force --kernels `rpm -q --queryformat '%{VERSION}-%{RELEASE}.%{ARCH}' kernel-devel`
```

Then modify /boot/loader/entries and add `rd.driver.blacklist=nouveau modprobe.blacklist=nouveau nvidia-drm.modeset=1` which will enable it on local machine only. So far I'm not sure if the boot entry update will survive bootc update.
